README.md
# ONTdeCIPHER

ONTdeCIPHER is an Oxford Nanopore Technology (ONT) amplicon-based sequencing pipeline to perform key downstream analyses on raw sequencing data from quality testing to SNPs effect to phylogenetic analysis. 
ONTdeCIPHER integrates 13 bioinformatics tools, including Seqkit, ARTIC bioinformatics tool, PycoQC, MultiQC, Minimap2, Medaka, Nanopolish, Pangolin (with the model database pangoLEARN), Deeptools (PlotCoverage, BamCoverage), Sniffles, MAFFT, RaxML and snpEff. 
The users can pre-process their data and obtain the sequencing statistics, reconstruct the consensus genome, identify variants and their effects for each viral isolate, infer lineage, and perform multi-sequence alignments and phylogenetic analyses. 
Currently, ONTdeCIPHER is mainly used to analyze the genetic diversity of SARS-CoV-2. However, any amplicon-based genome of pathogens can be analyzed if the primers scheme and a reference genome are available. 

<p align="center">
<img src="Images/ONTdeCIPHER_pipelineGit.png"
     width="50%"
     height="50%" 
     alt="deCIPHER"
     style="float: left; margin-right: 10px;" />
</p>

## Installation
### Requirements:
Python >=3\
Conda >=3

### 1. Downloading the source:
```sh
git clone https://github.com/emiracherif/ONTdeCIPHER
cd ONTdeCIPHER
```
### 2. Installing dependencies:

#### For Linux distributions
```sh
conda env create --name ontdecipher --file=Environments/ontdecipher_linux.yml
```
#### For macOS 
```sh
conda env create --name ontdecipher --file=Environments/ontdecipher_macOS.yml
```
For macOS users please check Tips section blow.
### 3. Installing artic & pangolin:
to install artic : 
```sh
conda activate ontdecipher
cd fieldbioinformatics && python setup.py install
```

to install pangolin : 
```sh
cd ../Pangolin && bash install_pangolin.sh
```
This step allows having the last version of pangolin and pangoLEARN.
## Usage
Before running ONTdeCIPHER, you will need first to create a working directory then put into it all your sequencing data (fastq/fastq.gz and fast5 files). ONTdeCIPHER will output the results in the working directory. 


### Structure of the working directory

	Working_Directory
		├── fastq_pass
		│   ├── barcode01
		│   ├── barcode02
		├── fast5_pass
		│   ├── barcode01
		│   ├── barcode02


### To run ONTdeCIPHER :

You have to run the master script `run_ONTdeCIPHER.py` from working directory by:
1) activating ONTdeCIPHER conda environment:

```sh
conda activate ontdecipher
```

2) running the master script

```sh
python3 absulote_path_to_script_directory/run_ONTdeCIPHER.py --step pip_core --params config.txt --samples config_samplename.tsv -t 10
```

`--step` : can be one of the following values: `pycoQC` , `pip_core` , `m_r_p` . To run all steps you can use : `all` .

`pycoQC` runs only PycoQC (computes metrics and generates interactive QC plots for Oxford Nanopore technologies sequencing data)
`pip_core`  runs ONTdeCIPHER core pipeline (artic, seqKit, DeepTools, snpEff)
`m_r_p` runs mafft, raxmlHPC and Pangolin
`all` runs `pycoQC` (if sequencing_summary.txt file is provided), `pip_core` & `m_r_p`


`--params` : a config file containing some parameters to run the pipeline.
#### Example:

	input_fastq="absulote_path_to_fastq_pass_directory/fastq_pass"
	input_fast5="absulote_path_to_fast5_pass_directory/fast5_pass"
	input_sequence_summary="path_to_sequencing_summary_file/sequencing_summary_xxx.txt"
	# minimum read length
	min="400"
	# maximum read length
	max="700" 
	# RaxML output name
	name="name_bootstrap"
	# SARS-CoV-2 reference in the snpEff data base.
	reference_genome_snpEff ="MN908947.3"
	# absulote path to sniffles
	sniffles="absulote_path_to_sniffles"
	

`--samples` : a config file to associate barcodes with sample names.
#### Example:

	#barcode<tab>sampleName
	barcode01	e1
	barcode02	e2
	barcode03	e3
	barcode04	e4
	barcode05	p1
	barcode06	p2
	barcode07	p3
	barcode08	p4

`--threads/-t` : Maximum number of threads to use. Default: 4

### Pipeline output results

After running ONTdeCIPHER steps you will hava in your working directory the following files and folders.


	├── DagFiles
	├── RAxML_bestTree.name_bootstrap
	├── RAxML_bipartitions.name_bootstrap
	├── RAxML_bipartitionsBranchLabels.name_bootstrap
	├── RAxML_bootstrap.name_bootstrap
	├── RAxML_info.name_bootstrap
	├── Step1_usedConfigs
	├── Step2_artic_guppyplex_filter
	├── Step3_artic_medaka_result
	├── Step4_artic_nanopolish_result
	├── Step5_snpEff_result
	├── Step6_plotCoverage_result
	├── Step7_bamCoverage_result
	├── Step8_consensus_fasta
	├── Summary
	├── fast5_pass
	├── fastq_pass
	├── global_lineage_information.csv
	├── lineage_report.csv
	└── report.html

Data to test ONTdeCIPHER and the results you will obtain are available here:

https://osf.io/jd2vz/?view_only=6d333ddc5a3045d297d5e3cc59e7e461
## Tips
**SnpEff**
 
SnpEff needs a database to perform genomic annotations. There are pre-built databases for thousands of genomes.
So to know which genomes have a pre-built database run (ONTdeCIPHER environment needs to be activated):
```sh
java -jar snpEff.jar databases
```
If your genome is absent from the database, you can build your database (see http://pcingola.github.io/SnpEff/se_buildingdb/).

**Sniffles**:

Last version of Sniffles available for macOS via conda is v1.0.7 (https://anaconda.org/bioconda/sniffles). We noticed that Sniffles v1.0.12 is more efficient for that we recommend macOS users to install the last version of Sniffles via git from : (https://github.com/fritzsedlazeck/Sniffles). To compile the sniffles correctly macOS user need to install `libomp` (via `brew` for example). After the installation, you need the path to libomp library to used instand of `/usr/local/Cellar/libomp/13.0.0` in commands blow (if it is different):

```sh
wget https://github.com/fritzsedlazeck/Sniffles/archive/master.tar.gz -O Sniffles.tar.gz
tar xzvf Sniffles.tar.gz
cd Sniffles-master/
mkdir -p build/
cd build/

cmake .. -DOpenMP_C_FLAGS=-fopenmp=lomp \
-DOpenMP_CXX_FLAGS=-fopenmp=lomp -DOpenMP_C_LIB_NAMES="libomp" \ 
-DOpenMP_CXX_LIB_NAMES="libomp" \
-DOpenMP_libomp_LIBRARY="/usr/local/Cellar/libomp/13.0.0/lib/libomp.dylib" \
-DOpenMP_CXX_FLAGS="-Xpreprocessor \
-fopenmp /usr/local/Cellar/libomp/13.0.0/lib/libomp.dylib \
-I/usr/local/Cellar/libomp/13.0.0/include" -DOpenMP_CXX_LIB_NAMES="libomp" \
-DOpenMP_C_FLAGS="-Xpreprocessor \
-fopenmp /usr/local/Cellar/libomp/13.0.0/lib/libomp.dylib \
-I/usr/local/Cellar/libomp/13.0.0/include"

cd ../bin/sniffles*
./sniffles

```
To use this version of sniffles, you can provid the absolute path to `absulote_path_to/Sniffles-master/bin/sniffles` in the `config.txt` file.


## References
### Pre-processing and quality control
Adrien Leger, Tommaso Leonardi, February 28, 2019, pycoQC, interactive quality control for Oxford Nanopore Sequencing
(https://tleonardi.github.io/pycoQC/)

Wei Shen, Shuai Le, Yan Li, Fuquan Hu, October 5, 2016, SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation
(https://bioinf.shenwei.me/seqkit/)  

Fidel Ramírez, Friederike Dündar, Sarah Diehl, Björn A. Grüning, Thomas Manke, May 05, 2014, deepTools: a flexible platform for exploring deep-sequencing data
(https://deeptools.readthedocs.io/en/develop/content/installation.html)

Ewels, P., Magnusson, M., Lundin, S., & Käller, M. (2016). MultiQC: summarize analysis results for multiple tools and samples in a single report
(https://multiqc.info/)

### Genome reconstruction and genomic analysis
Nick Loman, Andrew Rambaut, Jannuary 22, 2020, nCoV-2019 novel coronavirus bioinformatics environment setup 
(https://artic.network/ncov-2019/ncov2019-it-setup.html)

Li, H. (2018). Minimap2: pairwise alignment for nucleotide sequences
(https://github.com/lh3/minimap2)

Sedlazeck, F.J., Rescheneder, P., Smolka, M. et al. 2018.Accurate detection of complex structural variations using single-molecule sequencing
(https://github.com/fritzsedlazeck/Sniffles)

Pablo Cingolani et al, April 01, 2012, SnpEff: A program for annotating and predicting the effects of single nucleotide polymorphisms  
(http://pcingola.github.io/SnpEff/se_introduction/)

Katoh, Rozewicki, Yamada, 2019, MAFFT online service: multiple sequence alignment, interactive sequence choice and visualization    
(https://mafft.cbrc.jp/alignment/software/)

Alexandros Stamatakis, January 21, 2014, RAxML: a tool for phylogenetic analysis and post-analysis of large phylogenies 
(https://www.metagenomics.wiki/tools/phylogenetic-tree/construction/raxml) 

Huerta-Cepas, J., Serra, F., & Bork, P. 2016. ETE 3: reconstruction, analysis, and visualization of phylogenomic data
(http://etetoolkit.org/)

Andrew Rambaut et al, July 15 2020, A dynamic nomenclature proposal for SARS-CoV-2 lineages to assist genomic epidemiology 
(https://cov-lineages.org/pangolin.html)


## Project contributors

* Fatou Seck Thiam (BCD/ISEM/IRD), Marine Combe (ISEM/IRD),Georgina Rivera-Ingraham (ISEM/IRD), Fabienne Justy (ISEM/UM), Damien Breugnot (ISEM/IRD), Mohammad Salma (IGMM), Marie-ka Tilak (ISEM), Jean-Claude Doudou (IRD Cayenne), Rodolphe E. Gozlan (ISEM/IRD) and Emira Cherif (ISEM/IRD).
* **Written by Mohammad Salma, Fatou Seck Thiam and Emira Cherif.**


## Contact 
emira.cherif@ird.fr

## License
Licencied under CeCill-C (http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html) and GPLv3.                                                  

Intellectual property belongs to IRD and authors.

 



fieldbioinformatics/CHANGELOG
v1.1.0-rc1:
    * Support for read groups:
        * Support ‘pool’ read groups taken from BED file, e.g.:
            * nCoV2019_1
            * nCoV2019_2
        * These can be helpfully viewed in the Tablet viewer.
        * Do variant calling separately on each group, permits detection of incongruent mutations e.g. from contamination
    * Update to nanopolish 0.12.5 fo more informative VCF output
    * Change nanopore filtering to support new VCF output (e.g. strand-bias)
    * Support indels with bcftools consensus
    * Longshot added to Medaka to permit filter VCF on depth
    * New workflow model - all commands run "one per sample", script nanopolish index
    * Add 'artic gupplyplex' for support for demultiplexed Guppy output to replace 'artic gather'
    * Fix for align_trim to remove erroneous CIGAR strains that can cause medaka to fail
    * Support for amplicons v3 format file (Will Rowe)
    * Test suite (Will Rowe)
    * Documentation (Will Rowe)
    * Remove old deprecated code (Will Rowe)

v0.13:
    * ARTIC fork

v0.12: 
    * Update to nanopolish v0.8.4 for Albacore 2.0+ support:
        * run nanopolish -d /path/to/fast5 input.fasta before zibra.py minion

v0.11-dev:
    * Add stringent pipeline

v0.11 (25th July 2017):
    * Revert back to Ryan Wick's porechop repo as --untrimmed now implemented
    * New script: align_trim_fasta.py and reconstitute.py as basis for more principled demultiplexing (align first, then trim to known amplicon end points +/- 40 bases) to help prevent chimeric reads resulting in incorrect barcode identification

v0.10 (15th July 2017):
    * Better handling of confident deletions (previously ignored, now N-masked)
    * Update to nanopolish HEAD
    * Fixes to support latest nanopolish VCF format changes
    * Ability to choose higher values of max-haplotypes to support more divergent references

v0.9:
    * New command line interface zibra.py to ease extraction and demultiplexing with Albacore and multiple groups

v0.8:
    * Track latest Porechop with better adaptor trimming

v0.7:
    * Add Illumina wrapper scripts

v0.6:
    * Update poretools

v0.5:
    * Update nanopolish to latest HEAD - to fix Albacore 1.1 support (7de633d01cc35a58e5537af5dd1024ae0040d15c)
    * Add Lassa virus scheme

v0.4:
    * Update nanopolish to latest HEAD - aaaf90beb4efe37243c651e62fcfd11374fe2176 for Albacore 1 support
    * Initial support for Yellow Fever (500 and 1000 bp schemes)
        * Move to poretools for extraction with named basecaller

v0.3:
    * add Porechop

v0.2:
    * update nanopolish to 0.6 - adds default support for various nanopore models
    * update samtools to 1.4



Scripts/run_ONTdeCIPHER.py
#!/usr/bin/env python

#############################################################################################
# run_snakFile.py
#
# Copyright (C) 2021 Fatou Seck Thiam , Mohammad Salma & Emira Cherif
#
# This script is used to run the pipeline steps
#
# usage:
# python3 run_snakFile_v2.py [-h] --step STEP --params PARAMS --samples SAMPLES --threads INT
#############################################################################################
#############################################################################################

import os, sys, pwd, argparse # glob, re,
from datetime import datetime
import subprocess
from pathlib import Path
import pandas as pd
import numpy as np
import io, glob
import seaborn as sns
import matplotlib.pyplot as plt
import time
import resource
import tracemalloc
from colorama import Fore, Back, Style




###############################################################################
############# Some global variables ###########################################
## The list of steps
steps_list=["pycoQC","pip_core","m_r_p","all"]

## Config file exemple
config_file_exemple="""

			#barcode	sample_name
			barcode01	sample_1
			barcode02	sample_2
			barcode03	sample_3

			"Note: the columns are separated by '\t'."\n """

############# Functions's section #############################################

def getSnakeDir():

	""" Read the argv and extract the path to this script to use it later to
	call the 'covid.smk' file.
    Warning: the feature is expected to have sankemake.smk file in the same
    directory (i.e. no check
    is done).

    Return the path as a string """

	tool_dir="{}".format(sys.argv[0])
	directorys=tool_dir.split("/")[0:-1]
	tool_dir="/".join(directorys)
	return (tool_dir)


def getCondaPath():
	commandConda="echo $CONDA_PREFIX"
	cmd=subprocess.Popen(commandConda, shell=True, stdout=subprocess.PIPE)
	cmd_data=""
	for line in cmd.stdout:
		line= line.decode(encoding="utf-8", errors="ignore")
		cmd_data+=line
	print(cmd_data)
	return cmd_data.split("/envs/ontdecipher")[0]

def printInfo():

	""" Print on screen some usefull information like :
	the date of the run, the user who launched this script, the working
	directory, the dirctory of this script.

	Return NULL """

	print(Fore.GREEN +"\n -------------------------------",'\n',
	"This job has been started on :",datetime.today().strftime('%d-%m-%Y'),
	"at :", datetime.today().strftime('%H:%M:%S'),'\n',
	"-------------------------------",'\n',
	"Launched by:",pwd.getpwuid( os.getuid())[4],
	", with this username:",pwd.getpwuid( os.getuid())[0],'\n',
	"-------------------------------\n Working directory is:")

	subprocess.call(["pwd"])
	print(Fore.YELLOW +"-------------------------------",'\n',
		"You call this script form:", '\n',
		getSnakeDir(),'\n')



def readConfigFile(samplesf, paramsf, threads):

	""" Read the config file.

    Return sample name, params """

	getSnakeDir_path= '"'+str(getSnakeDir())+'"'
	conda_path='"'+str(getCondaPath())+'"'
	mySampleDict={}
	# Open the config file
	with open(samplesf, 'r') as infile:
			for line in infile:
				if not line.startswith('#'):
					try:
						# Split file lines by 'tab'
						columns = line.split("\t")
						# Check if the line dont start by '#'
						if (len(columns)!=2):
							print("ERROR: The ",samplesf," is not compatible. There are :",str(len(columns)),
								"Columns but 2 columns are expected"'\n',
						config_file_exemple)
							sys.exit(1)
						else:
							if columns[0] not in mySampleDict:
								mySampleDict.update({columns[0].strip(" "):columns[1].strip("\n")})
					except IndexError:
						print("ERROR: The ",samplesf," is not compatible. Please respect this structure",'\n',
						config_file_exemple)
						sys.exit(1)


	subprocess.call(["mkdir","-p","Step1_usedConfigs"])
	subprocess.call(["cp",samplesf,"Step1_usedConfigs/config_samplename.tsv"])
	if (paramsf):
		myParamDict={}
		subprocess.call(["mkdir","-p","Step1_usedConfigs"])
		subprocess.call(["cp",paramsf,"Step1_usedConfigs/config.txt"])
		cmd="""echo '\nscripts_path="""+getSnakeDir_path+"""' >> Step1_usedConfigs/config.txt"""
		subprocess.Popen(cmd,shell=True)
		cmd="""echo '\nconda_path="""+conda_path+"""' >> Step1_usedConfigs/config.txt"""
		subprocess.Popen(cmd,shell=True)
		cmd="""echo '\ncpu="""+str(threads)+"""' >> Step1_usedConfigs/config.txt"""
		subprocess.Popen(cmd,shell=True)

		subprocess.call(["sleep","3"])

		with open("Step1_usedConfigs/config.txt","r") as myConfig:
			for line in myConfig:
				if not line.startswith('#'):
					try:
						line=line.replace(" ", "").replace("\"", "").replace("'", "").replace("\t", "").strip("\n")
						if line!="\n":
							print(line)
							columns=line.split("=")
							if columns[0] not in myParamDict:
								myParamDict.update({columns[0]:columns[1].strip("\n")})
					except:
						pass
	#print(myParamDict)
	return(mySampleDict, myParamDict)



####################################################
def runPipeline(stepf,coref, mySampleDict, myParamDict):

	""" Run the called script. This function need 4 parameters :
	step : the name of the step to run.
	threads : the number of core.
	mySampleDict : Dict for file name in fast5/fastq directories and samples name.
	myParamDict : Dict for parameters.

	Return NULL """


	# Get snakefile path
	pathToSnake=getSnakeDir()
	###---------------------
	# Run pycoQC
	if (stepf==steps_list[0]):
		p = os.popen('pycoQC -f '+str(myParamDict['input_sequence_summary'])+' -o pycoQC_report.html')
		print(p.read())

	###---------------------
	# Run snakemake
	if (stepf==steps_list[1]):
		print("Running snakemake file ...")
		subprocess.call(["mkdir","-p","DagFiles"])

		snakeFileTarget="ontdeCIPHER.smk"
		if "input_fast5" in myParamDict.keys():
			snakeFileTarget="ontdeCIPHER_Q5.smk"

		p = os.popen('snakemake -s '+pathToSnake+'/'+snakeFileTarget+' --cores '+str(coref) +' --dag | dot -Tpdf > DagFiles/dag_start_'+str(datetime.today().strftime('at_%H-%M-%S_on_%d-%m-%Y'))+'.pdf')
		print(p.read())

		p = os.popen('snakemake -s '+pathToSnake+'/'+snakeFileTarget+' --cores '+str(coref))
		print(p.read())

		p = os.popen('snakemake --report report_'+str(datetime.today().strftime('at_%H-%M-%S_on_%d-%m-%Y'))+'.html -s '+pathToSnake+'/'+snakeFileTarget)
		print(p.read())

		p = os.popen('snakemake -s '+pathToSnake+'/'+snakeFileTarget+' --cores '+str(coref) +' --dag | dot -Tpdf > DagFiles/dag_end_'+str(datetime.today().strftime('at_%H-%M-%S_on_%d-%m-%Y'))+'.pdf')
		print(p.read())


	###---------------------
	# Run mafft, raxmlHPC & pangolin
	if (stepf==steps_list[2]):

		print("Running step2 ...")
		subprocess.call(["mkdir","-p","Step8_consensus_fasta"])
		p = os.popen('cat Step3_artic_medaka_result/*.consensus.fasta > Step8_consensus_fasta/all_fasta.fasta')
		print(p.read())

		p = os.popen('mafft --thread 4 --threadtb 5 --threadit 0 --reorder --auto Step8_consensus_fasta/all_fasta.fasta > Step8_consensus_fasta/all_alignment.fasta')
		print(p.read())

		p = os.popen('raxmlHPC -m GTRGAMMA -p 12345 -s Step8_consensus_fasta/all_alignment.fasta -n '+str(myParamDict['name'])+' -f a -x 1000 -N 100')
		print(p.read())


		cmd='source '+str(myParamDict['conda_path'])+'/etc/profile.d/conda.sh && conda activate ete3 && python3 '+pathToSnake+'/plot_tree.py --file RAxML_bestTree.'+str(myParamDict['name'])+' && conda deactivate'
		subprocess.Popen(cmd, shell=True, executable='/bin/bash')

		cmd='source '+str(myParamDict['conda_path'])+'/etc/profile.d/conda.sh && conda activate pangolin && pangolin Step8_consensus_fasta/all_fasta.fasta --threads '+str(coref)+' && conda deactivate'
		subprocess.Popen(cmd, shell=True, executable='/bin/bash')
		#print(p.read())

		####
		



		inputfiles=""
		for key, val in mySampleDict.items():
			inputfiles+="Step3_artic_medaka_result/"+val+"_"+key+".trimmed.rg.sorted.bam "

		print(inputfiles)
		p = os.popen("plotCoverage --bamfiles "+inputfiles+"-o Summary/All.pdf --smartLabels --plotWidth 25 --plotHeight 7 --outRawCounts Summary/All.txt --outCoverageMetrics Summary/All.txt --plotFileFormat pdf -p "+str(coref))
		print(p.read())


		####
		samples_dir = sorted(glob.glob("Summary/*_snake_Q.txt"))
		listSample=[]
		for samp in samples_dir:
			listSample.append(samp)
		#print(listSample)

		df_all="sample,file,format,type,num_seqs,sum_len,min_len,avg_len,max_len\n"
		for file in listSample:
			with open(file, 'r') as infile:
				nameSample=file.split("_snake1")[0].split("/")[-1]

				df = ""
				for line in infile:
					newLine=""
					if (not line.startswith('#')):
						if (not line.startswith('--')):
							if (not line.startswith('file')):
								columns =line.split(" ")
								for col in columns:
									if col != "":
										if col=="-":
											col="before_filtring"
										if col.startswith('Step2'):
											col ="after_filtring"

										newLine+= col+","

								#print(nameSample+","+newLine.strip(",")+"\n")
								df+=nameSample+","+newLine.strip(",")+"\n"
			df_all+=df

		###
		data = io.StringIO(df_all)
		df2 = pd.read_csv(data, sep=",")
		df2['sample'] = df2['sample'].str.replace("_snake_Q.txt","")
		df2['sample'] = df2['sample'].str.replace("_snake_QC.txt","")

		# set plot style: grey grid in the background:
		sns.set(style="darkgrid")

		# Set the figure size
		plt.figure(figsize=(20, 15))

		# grouped barplot
		sns.barplot(x="sample", y="num_seqs", hue="file", data=df2, ci=None);
		plt.xticks(rotation=45)
		plt.ylabel("number of sequences")
		plt.xlabel("Samples")
		plt.title("Summary")
		plt.savefig("Summary/stat.pdf", bbox_inches='tight')



		gb = df2.groupby("file")
		listTilte=["After","Before"]
		i=0
		for x in gb.groups:

			df3= pd.DataFrame(gb.get_group(x),columns=["sample","min_len","avg_len","max_len"])
			df3['sample'] = df3['sample'].str.replace("_snake_Q.txt","")
			df3['sample'] = df3['sample'].str.replace("_snake_QC.txt","")
			print (df3)
			df3 = df3.set_index('sample')
			plt.figure(figsize=(70, 75))
			df3.plot.barh(stacked=True, title= listTilte[i]);
			plt.xticks(rotation=45, fontsize=7)
			plt.yticks(rotation=45, fontsize=7)
			plt.xlabel("number of sequences")
			plt.ylabel("Samples")
			plt.legend(borderaxespad=0, fontsize=7)
			i+=1
			plt.savefig("Summary/summary_"+str(listTilte[i-1])+".pdf", bbox_inches='tight')


		print("Run multiqc")
		p = os.popen("multiqc . --dirs-depth 2")
		print(p.read())

	###---------------------
	# Run covid.smk file
	if (stepf==steps_list[3]):
		print("Running all the pipeline : pycoQC --> pip_core --> mafft, raxmlHPC and pangolin ;)")
		if "input_sequence_summary" in myParamDict.keys():
			runPipeline("pycoQC",coref,mySampleDict,myParamDict)
		runPipeline("pip_core",coref,mySampleDict,myParamDict)
		runPipeline("m_r_p",coref,mySampleDict,myParamDict)




####################################################
####################################################


def main():
	# Get some infos
	tracemalloc.start()

	time_start = time.perf_counter()
	printInfo()

	##############################################################################

	parser = argparse.ArgumentParser(description='run_ONTdeCIPHER.py')
	parser.add_argument('--step', help='Select step: pycoQC, pip_core, m_r_p or all. pycoQC: runs pycoQC tool. pip_core: runs ontdeCIPHER.smk. m_r_p: runs mafft, raxmlHPC and pangolin', required=True)
	parser.add_argument('--params', help='Config file containing some parameters to run the pipeline.', required=True)
	parser.add_argument('--samples', help='Config file to associate barcodes with sample names.', required=True)
	parser.add_argument('--threads','-t', type=int , default= 4, help='The number of cores to run the pipeline')
	args = parser.parse_args()

	mySampleDict, myParamDict = readConfigFile(args.samples,args.params,args.threads)

	if args.step in steps_list:
		runPipeline(args.step,args.threads, mySampleDict, myParamDict)

	else:
		print("ERROR: ",args.step," is not supported. Please select one these:",steps_list)
		sys.exit(1)

	####################################################
	print("###################")
	print(Fore.GREEN +"Completed at:",str(datetime.today().strftime('%H:%M:%S on %d-%m-%Y')))
	time_elapsed = (time.perf_counter() - time_start)
	dur=time.strftime('%H hour %M min %S sec', time.gmtime(time_elapsed))
	print(Fore.BLUE +"Duration: {} ".format(dur))

	# current, peak = tracemalloc.get_traced_memory()
	# print(Fore.BLUE + f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")

	tracemalloc.stop()

	print(" Done ! ")


if __name__ == '__main__':
	main()
sys.exit(0)
